### ISSUE TITLE
First experiments on `taller_puzzles` (mult vs single att / roll)

### ISSUE THREAD

**Issue by GnarlyMshtep** [GnarlyMshtep/pass_at_k#19]
**One sentence**
First experiment on `taller_puzzles` using multi_attempt/rollout

**What is interesting to measure?**
* comparison to single attempt in solve rate
* do the multiple attempts become diverse? 
* does the thinking display distinct thinking?
* correlation between thinking and n_vertices. 

**how will u measure it? (where will you read the #-s, which functions to add, etc)**
* wandb 
* script reading rollouts
* vibes 
* also external script. (print histogram)

(easier not to try to thread through verl program)

**What do you expect?**
* exceed single attempt
* yes, the model should never guess the same thing multiple times
* not sure. it could be that the model will concentrate all its thinking in the <think-1> and put placeholders in the latter. 
* linear? I think the actual complexity of solving the problem grow linearly because 3rd is fixed and small (n/2-th tallest would be a different story, might be NP hard). 

**Planned config**
* use dapo to avoid running into ceiling effects. See time for 4096+ 2048, if too long do 2048+ 2048. 
* use 2 GPUs for this run and 2 GPUs for a similar run but without multi-attempt. In that run I will just use DAPO and single-attempt. 
* For both runs I will cancel then "fill buffer with nonzero advantages feature" 

**Link to wandb and any other configs** 

**Results**

**Comment by GnarlyMshtep** [comment]
What is the baseline? Single attempt GRPO for the same problem

**Comment by GnarlyMshtep** [comment]
https://wandb.ai/stalaei-stanford-university/matan/runs/6k9nozdw?nw=nwusermatanshtepel

<img width="1308" height="362" alt="Image" src="https://github.com/user-attachments/assets/13813593-618b-494a-b88a-b1a34a59e670" />

both @ShayanTalaei  and I are seeing collapse. The model is not even trying 
```
<attempt-1>A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z</attempt-1>\n<attempt-2>A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z</attempt-2>\n<attempt-3>A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z</attempt-3>"

<think></think>\n<attempt-1>A</attempt-1>\n<attempt-2>B</attempt-2>\n<attempt-3>C</attempt-3>
```

### What should we try? 
I am trying the DAPO overlong filtering and trying to make the task harder

**Comment by GnarlyMshtep** [comment]
### Overlong 
response length still dying -- its not making reward totally die, even spiking up, so maybe... eh... 

<img width="1713" height="678" alt="Image" src="https://github.com/user-attachments/assets/34e46a14-dbe0-4674-8771-1efa61c42b97" />

<img width="1734" height="660" alt="Image" src="https://github.com/user-attachments/assets/bf50c061-3fd1-4b3b-b27a-f7a5fda5e6ae" />

**Comment by GnarlyMshtep** [comment]
### Harder 
min nodes in problem 4 --> 7. 
doesn't show that much of a difference (https://wandb.ai/stalaei-stanford-university/matan/runs/jnwuewly?nw=nwusermatanshtepel) -- how is it getting these problems without thinking?

Looking at rollouts its learning to get only the really easy ones. Let me kick off a run without the really easy ones.  

### Harder + Overlong + Larger n_rollout (stability) 
n_rollout 8 --> 16 -- also filter out where there is only one person who is third tallest. 

again collapse because it was basically never getting it right so it just guessed

<img width="1744" height="659" alt="Image" src="https://github.com/user-attachments/assets/0a957ad9-db67-4357-8284-d224806a2ac0" />

<img width="1735" height="358" alt="Image" src="https://github.com/user-attachments/assets/cc4e176b-2f79-4616-a35b-d1818f0c0f3a" />

**Comment by GnarlyMshtep** [comment]
<img width="1729" height="675" alt="Image" src="https://github.com/user-attachments/assets/d6af9d72-326d-4782-a086-78e7d94de90c" />
https://wandb.ai/stalaei-stanford-university/matan/runs/bomnev5d?nw=nwusermatanshtepel

still no. I think the problem is that the reward is too unstable -- it basically goes between 0 and 10% correct. I am interested in seeing the max per iteration. I think some possible solutions are 
* fine-tuning warmup
* larger batch size <-- **try first**
* larger n_rollout.

**Comment by GnarlyMshtep** [comment]
https://stanfordldrlab.slack.com/archives/C09BQGKR10R/p1756756495102559

7B LMs fail even the very easy puzzles -- (Qwen Instruct here)

**Comment by GnarlyMshtep** [comment]
We should discuss at meeting with @andrewilyas , but for now I am Closing becuase LMs are very incapable of this task and doesn't seem worth to continue investing in it? 


### the example that broke the camal's back

Even 7B LMs are having a very hard time with very simple height puzzles and that's why my setup is failing.
Here is my complete prompt including a puzzle on 4 nodes (basically trivial to solve)
<|im_start|>system\nYou are a helpful assistant who thinks step by step inside thinking tags and outputs guesses for the correct answer in attempt tags. You put ALL your thinking inside thinking tags. \\nYou put your attempts inside attempt tags. \\nYou ONLY put comma-separated lists of people inside attempt tags. \\nSo for example, if you decide to answer A,C,F for your second attempt, output <attempt-2>A,C,F</attempt-2> NOT <attempt-2>my second guess is A, C, and F</attempt-2> or something similar. The answer is always a comma-separated list with no spaces, so YOU ARE NOT ALLOWED TO PUT ENGLISH TEXT OR EXPLANATIONS INSIDE ATTEMPT TAGS<|im_end|>\n<|im_start|>user\nYou will be presented with a height comparison puzzle and you have 3 attempts to answer it correctly and you MUST think before each answer. So, your answer format must be <think></think> <attempt-1></attempt-1><attempt-2></attempt-2><attempt-3></attempt-3> where in <think> you think about the question and come up with 3 attempts. Generate your first attempt and consider why it might be wrong to generate your next attempts. Make sure to think about ALL you attempts in <think>. DON'T think only about the first attempt and then guess the rest. The attempt answers provided in <attempt-1>, <attempt-2>, and <attempt-3>. Please optimize for getting at least one attempt correct, rather than getting more than one attempt correct (pass@k grading). Try justifying your answer to yourself. Think hard about the answer, spending a long time contemplating. Your answer should be a comma-separated list of people who could plausibly be the 3rd tallest.\\n\\nA and B are the same height.\\nC is shorter than A\\nS is shorter than B<|im_end|>\n<|im_start|>assistant\n
correct answer is {C,S }
and here is a somewhat typical response
Given:
1. A and B are the same height.
2. C is shorter than A.
3. S is shorter than B.

From these statements, we can infer:
- Since A and B are the same height, they cannot be the third tallest.
- C is shorter than both A and B, so C cannot be the third tallest either.
- S is shorter than B, which means S also cannot be the third tallest.

This leaves us with D as the only person who could potentially be the third tallest. However, since there's no information about D's height relative to others, we need to consider other possibilities as well.

Possible candidates for the third tallest could be E or F, given that their heights aren't directly compared to A, B, C, or S. Therefore, my first attempt would be E, F, D.
</think>
<attempt-1>E,F,D</attempt-1>
<think>
Let's re-evaluate our choices. We know:
- A = B
- C < A
- S < B

If we assume D is not the third tallest, then the third tallest must be between E and F. But without direct comparisons involving E and F against everyone else, we need to ensure we cover all bases.

Since C and S are definitely shorter, and A and B are equal, the third tallest has to be someone taller than C but shorter than A/B. This means the third tallest could indeed be E or F, but not necessarily D unless we have more information. 

Thus, revisiting, I believe the most plausible options are still E and F, as D doesn't fit due to lack of comparative data.
</think>
<attempt-2>E,F</attempt-2>
<think>
Final check: 
- E and F are the only ones left after eliminating A, B, C, and S.
- If E and F are not the third tallest, then D would need to be considered again, but we don't have enough info on D.

Therefore, the best guess remains E and F, as they are the only ones who haven't been definitively eliminated based on the given information.
</think>
<attempt-3>E,F</attempt-3>
Not super sure what to do -- can try to train on these puzzles or prompt engineer even harder, but this is about as simple as these puzzles will go.
Was thinking about doing some reward shaping -- having a partial reward for being close to the right set -- but I think that at this level of capability that might not matter.
